be/src/runtime/runtime_filter_cache.cpp

```diff
@@ -0,0 +1,217 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2021-present, StarRocks Limited.
+
+#include "runtime/runtime_filter_cache.h"
+
+#include <chrono>
+
+#include "gutil/strings/substitute.h"
+#include "util/thread.h"
+
+namespace starrocks {
+
+using std::chrono::seconds;
+using std::chrono::milliseconds;
+using std::chrono::steady_clock;
+using std::chrono::duration_cast;
+class RfCacheValue {
+public:
+    RfCacheValue() { _extend_lifetime(); }
+    ~RfCacheValue() = default;
+    void put_if_absent(int filter_id, const JoinRuntimeFilterPtr& filter) {
+        _extend_lifetime();
+        if (!_filters.count(filter_id)) {
+            _filters[filter_id] = filter;
+        }
+    }
+
+    JoinRuntimeFilterPtr get(int filter_id) const {
+        _extend_lifetime();
+        auto it = _filters.find(filter_id);
+        if (it != _filters.end()) {
+            return it->second;
+        } else {
+            return nullptr;
+        }
+    }
+
+    bool is_expired() const {
+        auto now = duration_cast<milliseconds>(steady_clock::now().time_since_epoch()).count();
+        return now > _deadline;
+    }
+
+private:
+    void _extend_lifetime() const {
+        _deadline = duration_cast<milliseconds>(steady_clock::now().time_since_epoch() + EXPIRE_SECONDS).count();
+    }
+    std::unordered_map<int, std::shared_ptr<vectorized::JoinRuntimeFilter>> _filters;
+    mutable int64_t _deadline;
+    static constexpr seconds EXPIRE_SECONDS = seconds(60);
+};
+
+class RfEvent {
+public:
+    RfEvent(const TUniqueId& query_id, int filter_id, std::string&& msg)
+            : query_id(query_id),
+              filter_id(filter_id),
+              ts(duration_cast<milliseconds>(steady_clock::now().time_since_epoch()).count()),
+              msg(std::move(msg)) {}
+    ~RfEvent() = default;
+    std::string to_string() { return strings::Substitute("[$0][$1] $2: $3", print_id(query_id), filter_id, ts, msg); }
+
+private:
+    TUniqueId query_id;
+    int filter_id;
+    long ts;
+    std::string msg;
+};
+using RfEventPtr = std::shared_ptr<RfEvent>;
+using RfEvents = std::list<RfEventPtr>;
+class RfEventValue {
+public:
+    RfEventValue() = default;
+    ~RfEventValue() = default;
+    void add_event(const TUniqueId& query_id, int filter_id, std::string&& msg) {
```

Comment:
What's the meaning of this method?  this method is called in many places, which seems error-prone.

Reply:
add a runtime filter trace event. it is safe, only when we debug runtime filter, trace is turned on by restful api